ulid/_util.ts

```diff
@@ -0,0 +1,111 @@
+
+interface ULID {
+  (seedTime?: number): string;
+}
+
+interface PRNG {
+  (): number;
+}
+
+// These values should NEVER change. If
+// they do, we're no longer making ulids!
+export const ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"; // Crockford's Base32
+export const ENCODING_LEN = ENCODING.length;
+export const TIME_MAX = Math.pow(2, 48) - 1;
+export const TIME_LEN = 10;
+export const RANDOM_LEN = 16;
+
+function replaceCharAt(str: string, index: number, char: string) {
+  return str.substring(0, index) + char + str.substring(index + 1);
+}
+
+export function encodeTime(now: number, len: number = TIME_LEN): string {
+  if (now > TIME_MAX) {
+    throw new Error("cannot encode time greater than " + TIME_MAX);
+  }
+  if (now < 0) {
+    throw new Error("time must be positive");
+  }
+  if (Number.isInteger(now) === false) {
+    throw new Error("time must be an integer");
+  }
+  let str = "";
+  for (; len > 0; len--) {
+    const mod = now % ENCODING_LEN;
+    str = ENCODING[mod] + str;
+    now = (now - mod) / ENCODING_LEN;
+  }
+  return str;
+}
+
+export function encodeRandom(len: number): string {
+  let str = "";
+  const randomBytes = crypto.getRandomValues(new Uint8Array(len));
+  for (let i = 0; i < len; i++) {
+    str += ENCODING[randomBytes[i] % ENCODING_LEN];
+  }
+  return str;
+}
+
+export function incrementBase32(str: string): string {
+  let index = str.length;
+  let char;
+  let charIndex;
+  const maxCharIndex = ENCODING_LEN - 1;
+  while (index-- >= 0) {
+    char = str[index];
+    charIndex = ENCODING.indexOf(char);
+    if (charIndex === -1) {
+      throw new Error("incorrectly encoded string");
+    }
+    if (charIndex === maxCharIndex) {
+      str = replaceCharAt(str, index, ENCODING[0]);
+      continue;
+    }
+    return replaceCharAt(str, index, ENCODING[charIndex + 1]);
+  }
+  throw new Error("cannot increment this string");
+}
+
+/**
+ * Generates a ULID function
+ */
+export function factory(): ULID {
+  return function ulid(seedTime: number = Date.now()): string {
+    return encodeTime(seedTime, TIME_LEN) + encodeRandom(RANDOM_LEN);
+  };
+}
+
+/**
+ * Generates a monotonically increasing ULID, optionally given a PRNG.
+ *
+ * @example To generate monotonically increasing ULIDs, create a monotonic counter.
+ * ```ts
+ * import { monotonicFactory } from "https://deno.land/std@$STD_VERSION/ulid/_util.ts";
+ *
+ * const ulid = monotonicFactory();
+ * // Strict ordering for the same timestamp, by incrementing the least-significant random bit by 1
+ * ulid(150000); // 000XAL6S41ACTAV9WEVGEMMVR8
+ * ulid(150000); // 000XAL6S41ACTAV9WEVGEMMVR9
+ * ulid(150000); // 000XAL6S41ACTAV9WEVGEMMVRA
+ * ulid(150000); // 000XAL6S41ACTAV9WEVGEMMVRB
+ * ulid(150000); // 000XAL6S41ACTAV9WEVGEMMVRC
+ *
+ * // Even if a lower timestamp is passed (or generated), it will preserve sort order
+ * ulid(100000); // 000XAL6S41ACTAV9WEVGEMMVRD
+ * ```
+ */
+export function monotonicFactory(encodeRand = encodeRandom): ULID {
```

Comment:
How come this is here instead of `mod.ts`?

Reply:
See above.